### DISCUSSION TITLE
Open Track Selector by default; assembly-specific default tracks?

### DISCUSSION THREAD

**Discussion by robinohm** [GMOD/jbrowse-components#4065] [General]
I have two questions, and I hope I didn't overlook it in the documentation.

Is there a way to open the Track Selector panel by default  (i.e., when a user opens that assembly in JBrowse2)? I tried adding the following to the config.json file, but as far as I can tell it had no effect:
  "defaultSession": {"activeWidgets": {"hierarchicalTrackSelector": "hierarchicalTrackSelector"}},
  "session": {"activeWidgets": {"hierarchicalTrackSelector": "hierarchicalTrackSelector"}},

Another question: can I specify tracks that should be shown by default for each assembly? I'm setting up a JBrowse instance with 100+ assemblies, and the track names vary per assembly. It would be useful if I can specify the default tracks for each assembly somewhere in the config.json file. Currently I'm trying to solve it by using the URL parameters, but that is only a partial fix, since if the user opens a new Linear Genome View then it doesn't display any tracks.

Any help is greatly appreciated!

**Comment by cmdcolin** [comment]
@robinohm hi there

a) there is a way to do this with defaultSession, our volvox config.json has this. however, a defaultSession is just what is initially shown. to do it dynamically any time a view is opened, is another thing which we dont have
b) this is also not available dynamically any time a view is opened

both (a) and (b) are interesting ideas and could be considered for features. note that alternatively, a small plugin can recreate this funcionality

here is a plugin that does both (a) and (b)

```
;(function () {
  class Plugin {
    name = 'ShowTracksPlugin'
    version = '1.0'

    install(pluginManager) {
      const { when } = pluginManager.jbrequire('mobx') // get the "when" function from the mobx library

      pluginManager.addToExtensionPoint('Core-extendPluggableElement', elt => {
        if (elt.name === 'LinearGenomeView') {
          elt.stateModel = elt.stateModel.actions(self => ({
            afterCreate() {
              ;(async () => {
                await when(() => self.initialized) // wait for initialized, generally just good to do this but might not be strictly required
                self.activateTrackSelector() // (a) open track selector after the linear genome view is created (we are in the afterCreate callback)
                if(self.assemblyNames.includes('hg19')) { //e.g. if you have multiple assemblies, then check for e.g. hg19 before showing hg19 tracks....
                  self.showTrack('hg19_genes') // (b) these are trackId from the config.json file
                  self.showTrack('hg19_bigwig')// (b)
                } 
              })()
            },
          }))
        }
        return elt
      })
    }

    configure(/* pluginManager */) {}
  }

  // the plugin will be included in both the main thread and web worker, so
  // install plugin to either window or self (webworker global scope)
  ;(typeof self !== 'undefined' ? self : window).JBrowsePluginUMDLocPlugin = {
    default: Plugin,
  }
})()
```

**Reply by cmdcolin** [reply]
to use this plugin, copy it to a file in your jbrowse folder, and then use one of these options: https://github.com/GMOD/jbrowse-components/blob/ea6f1db592ece7daa0a5211f0ced515c9281c9da/test_data/volvox/config.json#L2170-L2181

you can use umdUrl:"plugin.js" which is relative to the index.html of the jbrowse, and umdLoc:{uri:"plugin.js"} is relative to the config.json (handy if you have multiple configs)

**Reply by cmdcolin** [reply]
small side note: 

this could have some consequences during use with synteny views or views where linear genome view is a subview e.g. breakpoint split view. 

the synteny view launches multiple linear genome views and so activateTrackSelector would be called by each of them, which would probably just result in one randomly being shown. might not be that bad. alternatively, you could additionally check if the lineargenomeview is e.g. inside a synteny view before doing the activateTrackSelector. 

and with things like breakpoint split view, maybe you dont want those 'default tracks; shown....you might only want the alignments evidence by default'....so then you conditionally only use the showTrack when its not a subview

these are just the types of considerations that are necessary before making a feature like this included in core jbrowse
